### ISSUE TITLE
OnComplete sometimes not called

### ISSUE THREAD

**Issue by tankin** [Demigiant/dotween#333]
code:
```
transform.DOLocalMove(originLocalPos, 0.2f)
           .OnComplete(() => {
               Debug.Log("I'm Called In Complete");
               StartCoroutine(DoSomething());
            });
```
It's weird that sometimes the function in OnComplete() is not called.
This happens about 20%.
Did I write wrong code?
shouldn't I write StartCoroutine in OnComplete?
I'm curious about this thing for a long time, though I can fix it by throw the startCorontine in another function, and call it in OnComplete, why this happens and I can't do that?

I use unity 2017.3.1p4(64bit)
mac mini
Dotween 1.2.250

Thanks for any userful information

**Comment by IvanMurzak** [comment]
Do you have DOTween Pro? I found a problem with DOTweenAnimation component. Maybe it is your case: https://github.com/Demigiant/dotween/issues/313#issuecomment-532184121

**Comment by tankin** [comment]
> 
> 
> Do you have DOTween Pro? I found a problem with DOTweenAnimation component. Maybe it is your case: [#313 (comment)](https://github.com/Demigiant/dotween/issues/313#issuecomment-532184121)

Thanks, but it's not that problem.

**Comment by irakli** [comment]
I'm having the exact same problem.

Unity 2019.2.5f1
DoTween 1.2.250 (free)

**Comment by Demigiant** [comment]
Unless you're encountering some weird fringe case, OnComplete is a very solid element of DOTween. Is it possible that you have something else intervening there and preventing the tween from completing? Or that you're overriding that OnComplete by assigning another, possibly NULL, one?

**Comment by tankin** [comment]
> Unless you're encountering some weird fringe case, OnComplete is a very solid element of DOTween. Is it possible that you have something else intervening there and preventing the tween from completing? Or that you're overriding that OnComplete by assigning another, possibly NULL, one?

I‘m sure It's a very simple program that there's no code to change the tween from complete. 
You can have a try for that.
Thank you all the same.

**Comment by Demigiant** [comment]
I did try with this code but it works perfectly:
```
    public Transform target;
    Tween t;

    void Start()
    {
        t = target.DOLocalMove(new Vector3(2, 2, 2), 0.2f)
            .OnComplete(() => {
                Debug.Log("I'm Called In Complete");
                StartCoroutine(CO_Do());
            });
    }

    IEnumerator CO_Do()
    {
        yield return null;
        t = target.DOLocalMove(new Vector3(2, 2, 2), 0.2f)
            .OnComplete(() => {
                Debug.Log("I'm Called In Complete");
                StartCoroutine(CO_Do());
            });
    }
```

Could you reproduce your issue in a MonoBehaviour script and attach it here so I can replicate it?

**Comment by j4nw** [comment]
I randomly started having a similar issue with particular `OnComplete`s and `Sequence`s with `AppendCallback`s often, but not always, just not firing. The sequence seems not to run at all, or just for a single frame of the first tween, and then errors are thrown a bit later, probably when new tweens are created.

Using Unity 2019.2.5f1 (or 6f1, same issue) and DOTween 1.2.280

I'll try to reproduce in a minimal example, but for now:

Here, a number of coin prefabs are instantiated in a loop, and given separate sequences. The very first coin instantiated at `i = 0`, and only this one, is almost always frozen in place, all the stuff just doesn't execute.
```
// some direction of "bursting into coins" is given 
var reward = rewards.Roll(); // int
for (var i = 0; i < reward; i++)
{
    var coin = Instantiate(coinPrefab, UI.instance.transform).GetComponent<GoldCoin>();
    coin.transform.position = self.transform.position;
    coin.transform.localScale = Vector3.one;

    var target = direction.ToVector2().normalized + Random.insideUnitCircle * 0.5f;
    var worldTarget = self.transform.position + target.ToVector3();
    var sequence = DOTween.Sequence();
    sequence.Append(coin.transform.DOJump(worldTarget, 1, 2, Random.Range(0.5f, 1f)).SetEase(Ease.Linear));
    sequence.AppendCallback(coin.Gather);
}

// coin.Gather()
transform.SetParent(UI.instance.transform);
transform
    .DOMove(UI.instance.loot.bagText.transform.position, 0.5f)
    .SetEase(Ease.InBack)
    .OnComplete(() =>
    {
        Player.instance.ChangeGold(1); 
        Destroy(gameObject);
    });
```

Here's something simpler where this happens, without any of that tween nesting in callbacks:
```
var sequence = DOTween.Sequence();
sequence.Append(piece.transform.DOScale(0, 0.2f));
sequence.AppendCallback(() => piece.sprite = newPiece.sprite);
sequence.AppendCallback(() => tooltip.Build()); // just a bunch of calls to a StringBuilder, nothing fancy
sequence.Append(piece.transform.DOScale(1, 0.2f));

background.DOColor(flashColor, 0.4f).SetLoops(2, LoopType.Yoyo);
```
The `background.DOColor` always runs, the sequence sometimes doesn't.

Turning off autoplay and explicitly playing the sequences doesn't change anything.

Errors with Safe Mode turned off for the first piece of code:
![dg_errors](https://user-images.githubusercontent.com/26443707/66028737-ff253c80-e4fd-11e9-9c39-a3dc19ec76c6.png)
```
MissingReferenceException: The object of type 'RectTransform' has been destroyed but you are still trying to access it.
Your script should either check if it is null or you should not destroy the object.
UnityEngine.Transform.set_localScale (UnityEngine.Vector3 value) <0x25db36c1d50 + 0x0005a> in <e3232360d51c46c786dd547e2f0b3f8b>:0
DG.Tweening.ShortcutExtensions+<>c__DisplayClass45_0.<DOScale>b__1 (UnityEngine.Vector3 x) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/ShortcutExtensions.cs:628)
DG.Tweening.Plugins.Vector3Plugin.EvaluateAndApply (DG.Tweening.Plugins.Options.VectorOptions options, DG.Tweening.Tween t, System.Boolean isRelative, DG.Tweening.Core.DOGetter`1[T] getter, DG.Tweening.Core.DOSetter`1[T] setter, System.Single elapsed, UnityEngine.Vector3 startValue, UnityEngine.Vector3 changeValue, System.Single duration, System.Boolean usingInversePosition, DG.Tweening.Core.Enums.UpdateNotice updateNotice) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Plugins/Vector3Plugin.cs:151)
DG.Tweening.Core.TweenerCore`3[T1,T2,TPlugOptions].ApplyTween (System.Single prevPosition, System.Int32 prevCompletedLoops, System.Int32 newCompletedSteps, System.Boolean useInversePosition, DG.Tweening.Core.Enums.UpdateMode updateMode, DG.Tweening.Core.Enums.UpdateNotice updateNotice) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenerCore.cs:247)
DG.Tweening.Tween.DoGoto (DG.Tweening.Tween t, System.Single toPosition, System.Int32 toCompletedLoops, DG.Tweening.Core.Enums.UpdateMode updateMode) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Tween.cs:252)
DG.Tweening.Core.TweenManager.Goto (DG.Tweening.Tween t, System.Single to, System.Boolean andPlay, DG.Tweening.Core.Enums.UpdateMode updateMode) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:706)
DG.Tweening.Sequence.ApplyInternalCycle (DG.Tweening.Sequence s, System.Single fromPos, System.Single toPos, DG.Tweening.Core.Enums.UpdateMode updateMode, System.Boolean useInverse, System.Boolean prevPosIsInverse, System.Boolean multiCycleStep) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Sequence.cs:331)
DG.Tweening.Sequence.DoApplyTween (DG.Tweening.Sequence s, System.Single prevPosition, System.Int32 prevCompletedLoops, System.Int32 newCompletedSteps, System.Boolean useInversePosition, DG.Tweening.Core.Enums.UpdateMode updateMode) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Sequence.cs:239)
DG.Tweening.Sequence.ApplyTween (System.Single prevPosition, System.Int32 prevCompletedLoops, System.Int32 newCompletedSteps, System.Boolean useInversePosition, DG.Tweening.Core.Enums.UpdateMode updateMode, DG.Tweening.Core.Enums.UpdateNotice updateNotice) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Sequence.cs:143)
DG.Tweening.Tween.DoGoto (DG.Tweening.Tween t, System.Single toPosition, System.Int32 toCompletedLoops, DG.Tweening.Core.Enums.UpdateMode updateMode) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Tween.cs:252)
DG.Tweening.Core.TweenManager.Update (DG.Tweening.UpdateType updateType, System.Single deltaTime, System.Single independentTime) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:490)
DG.Tweening.Core.DOTweenComponent.Update () (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/DOTweenComponent.cs:73)

MissingReferenceException: The object of type 'RectTransform' has been destroyed but you are still trying to access it.
Your script should either check if it is null or you should not destroy the object.
UnityEngine.Transform.set_localScale (UnityEngine.Vector3 value) <0x25db36c1d50 + 0x0005a> in <e3232360d51c46c786dd547e2f0b3f8b>:0
DG.Tweening.ShortcutExtensions+<>c__DisplayClass45_0.<DOScale>b__1 (UnityEngine.Vector3 x) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/ShortcutExtensions.cs:628)
DG.Tweening.Plugins.Vector3Plugin.EvaluateAndApply (DG.Tweening.Plugins.Options.VectorOptions options, DG.Tweening.Tween t, System.Boolean isRelative, DG.Tweening.Core.DOGetter`1[T] getter, DG.Tweening.Core.DOSetter`1[T] setter, System.Single elapsed, UnityEngine.Vector3 startValue, UnityEngine.Vector3 changeValue, System.Single duration, System.Boolean usingInversePosition, DG.Tweening.Core.Enums.UpdateNotice updateNotice) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Plugins/Vector3Plugin.cs:151)
DG.Tweening.Core.TweenerCore`3[T1,T2,TPlugOptions].ApplyTween (System.Single prevPosition, System.Int32 prevCompletedLoops, System.Int32 newCompletedSteps, System.Boolean useInversePosition, DG.Tweening.Core.Enums.UpdateMode updateMode, DG.Tweening.Core.Enums.UpdateNotice updateNotice) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenerCore.cs:247)
DG.Tweening.Tween.DoGoto (DG.Tweening.Tween t, System.Single toPosition, System.Int32 toCompletedLoops, DG.Tweening.Core.Enums.UpdateMode updateMode) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Tween.cs:252)
DG.Tweening.Core.TweenManager.Goto (DG.Tweening.Tween t, System.Single to, System.Boolean andPlay, DG.Tweening.Core.Enums.UpdateMode updateMode) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:706)
DG.Tweening.Sequence.ApplyInternalCycle (DG.Tweening.Sequence s, System.Single fromPos, System.Single toPos, DG.Tweening.Core.Enums.UpdateMode updateMode, System.Boolean useInverse, System.Boolean prevPosIsInverse, System.Boolean multiCycleStep) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Sequence.cs:331)
DG.Tweening.Sequence.DoApplyTween (DG.Tweening.Sequence s, System.Single prevPosition, System.Int32 prevCompletedLoops, System.Int32 newCompletedSteps, System.Boolean useInversePosition, DG.Tweening.Core.Enums.UpdateMode updateMode) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Sequence.cs:216)
DG.Tweening.Sequence.ApplyTween (System.Single prevPosition, System.Int32 prevCompletedLoops, System.Int32 newCompletedSteps, System.Boolean useInversePosition, DG.Tweening.Core.Enums.UpdateMode updateMode, DG.Tweening.Core.Enums.UpdateNotice updateNotice) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Sequence.cs:143)
DG.Tweening.Tween.DoGoto (DG.Tweening.Tween t, System.Single toPosition, System.Int32 toCompletedLoops, DG.Tweening.Core.Enums.UpdateMode updateMode) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Tween.cs:252)
DG.Tweening.Core.TweenManager.Update (DG.Tweening.UpdateType updateType, System.Single deltaTime, System.Single independentTime) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:490)
DG.Tweening.Core.DOTweenComponent.Update () (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/DOTweenComponent.cs:73)

MissingReferenceException: The object of type 'Image' has been destroyed but you are still trying to access it.
Your script should either check if it is null or you should not destroy the object.
UnityEngine.EventSystems.UIBehaviour.IsActive () (at C:/Program Files/Unity/Hub/Editor/2019.2.6f1/Editor/Data/Resources/PackageManager/BuiltInPackages/com.unity.ugui/Runtime/EventSystem/UIBehaviour.cs:28)
UnityEngine.UI.Graphic.SetVerticesDirty () (at C:/Program Files/Unity/Hub/Editor/2019.2.6f1/Editor/Data/Resources/PackageManager/BuiltInPackages/com.unity.ugui/Runtime/UI/Core/Graphic.cs:249)
UnityEngine.UI.Graphic.set_color (UnityEngine.Color value) (at C:/Program Files/Unity/Hub/Editor/2019.2.6f1/Editor/Data/Resources/PackageManager/BuiltInPackages/com.unity.ugui/Runtime/UI/Core/Graphic.cs:152)
DG.Tweening.DOTweenModuleUI+<>c__DisplayClass3_0.<DOColor>b__1 (UnityEngine.Color x) (at Assets/Demigiant/DOTween/Modules/DOTweenModuleUI.cs:64)
DG.Tweening.Plugins.ColorPlugin.EvaluateAndApply (DG.Tweening.Plugins.Options.ColorOptions options, DG.Tweening.Tween t, System.Boolean isRelative, DG.Tweening.Core.DOGetter`1[T] getter, DG.Tweening.Core.DOSetter`1[T] setter, System.Single elapsed, UnityEngine.Color startValue, UnityEngine.Color changeValue, System.Single duration, System.Boolean usingInversePosition, DG.Tweening.Core.Enums.UpdateNotice updateNotice) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Plugins/ColorPlugin.cs:79)
DG.Tweening.Core.TweenerCore`3[T1,T2,TPlugOptions].ApplyTween (System.Single prevPosition, System.Int32 prevCompletedLoops, System.Int32 newCompletedSteps, System.Boolean useInversePosition, DG.Tweening.Core.Enums.UpdateMode updateMode, DG.Tweening.Core.Enums.UpdateNotice updateNotice) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenerCore.cs:247)
DG.Tweening.Tween.DoGoto (DG.Tweening.Tween t, System.Single toPosition, System.Int32 toCompletedLoops, DG.Tweening.Core.Enums.UpdateMode updateMode) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Tween.cs:252)
DG.Tweening.Core.TweenManager.Update (DG.Tweening.UpdateType updateType, System.Single deltaTime, System.Single independentTime) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:490)
DG.Tweening.Core.DOTweenComponent.Update () (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/DOTweenComponent.cs:73)

IndexOutOfRangeException: Index was outside the bounds of the array.
DG.Tweening.Core.TweenManager.RemoveActiveTween (DG.Tweening.Tween t) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:1072)
DG.Tweening.Core.TweenManager.Despawn (DG.Tweening.Tween t, System.Boolean modifyActiveLists) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:236)
DG.Tweening.Core.TweenManager.DespawnActiveTweens (System.Collections.Generic.List`1[T] tweens) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:1058)
DG.Tweening.Core.TweenManager.Update (DG.Tweening.UpdateType updateType, System.Single deltaTime, System.Single independentTime) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:502)
DG.Tweening.Core.DOTweenComponent.Update () (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/DOTweenComponent.cs:73)

IndexOutOfRangeException: Index was outside the bounds of the array.
(wrapper stelemref) System.Object.virt_stelemref_class_small_idepth(intptr,object)
DG.Tweening.Core.TweenManager.ReorganizeActiveTweens () (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:1048)
DG.Tweening.Core.TweenManager.Update (DG.Tweening.UpdateType updateType, System.Single deltaTime, System.Single independentTime) (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/TweenManager.cs:402)
DG.Tweening.Core.DOTweenComponent.Update () (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTween/Core/DOTweenComponent.cs:73)
```

**Comment by j4nw** [comment]
Downgraded to v1.2.245 and it hasn't happened again in 30 minutes of testing.

**Comment by tankin** [comment]
> I did try with this code but it works perfectly:
> 
> ```
>     public Transform target;
>     Tween t;
> 
>     void Start()
>     {
>         t = target.DOLocalMove(new Vector3(2, 2, 2), 0.2f)
>             .OnComplete(() => {
>                 Debug.Log("I'm Called In Complete");
>                 StartCoroutine(CO_Do());
>             });
>     }
> 
>     IEnumerator CO_Do()
>     {
>         yield return null;
>         t = target.DOLocalMove(new Vector3(2, 2, 2), 0.2f)
>             .OnComplete(() => {
>                 Debug.Log("I'm Called In Complete");
>                 StartCoroutine(CO_Do());
>             });
>     }
> ```
> 
> Could you reproduce your issue in a MonoBehaviour script and attach it here so I can replicate it?

I'm sorry, recently I upgrade my Unity edition, now is 2018.3, so I cann't show the bug anymore
The problem focus on the coroutine maybe...

**Comment by aholla** [comment]
hi im getting a similar error, was this ever resolved?


Version 1.2.4xx and unity 2021.1.15

**Comment by ghost** [comment]
We had a similar problem and it was caused by one of our scripts setting Time.timeScale to 0 for awhile.

**Comment by Jan-Kiepski** [comment]
Hey, I was also struggling with this issue, and I've just figured out that it was caused by improper use of DOTween.KillAll(). For some reason, I thought it would kill all tweens that were started in the same script, but instead it kills ALL tweens that are currently in progress. So it makes sense that OnComplete wasn't called, because the tweens were killed before completing.

Now I use .SetID(this.gameobject) in tweens and later DOTween.Kill(this.gameObject) for a quick way to achieve behavior I wanted initially. Note: if you choose to do this, avoid attaching multiple "tween controller" components to the same object.
